### ISSUE TITLE
`@gen_cluster` has become 1s slower; CI takes 50% longer

### ISSUE THREAD

**Issue by crusaderky** [dask/distributed#6632]
Since #6603, a test decorated with `@gen_cluster`:
```python
@gen_cluster()
async def test1(s, a, b):
    pass
```
has increased in runtime on my host from 220ms to 1220ms.

With the same PR, the CI test suites (ci1) have increased:
Ubuntu: 19m23s -> 34m18s
Windows: 23m36s -> 48m55s
MacOSX: 32m15s -> 46m39s

CC @graingert

**Comment by hendrikmakait** [comment]
I'd be open to dropping the grace period and directly cancelling tasks. Are there any tasks for which we'd actually prefer graceful shutdown? In that case we could adjust the signature of `call_soon` to include a `graceful=False` keyword that would get stored together with the task itself and we could wait only for graceful shutdown of those tasks (likely with a shorter grace period).

**Comment by fjetter** [comment]
>  remove_worker_from_events should be called immediately when a worker reports that it is closing.

1. We need to remove workers from events, eventually. Depending on the deployment, this can cause a memory leak otherwise. I introduced this ~2 years ago because we were using clusters that were always-on but we scaled them up and down based on demand, i.e. over time this events dict became huge. Eventually this dict should be cleaned up. The solution back then was to introduce a timeout
2. It should be delayed such that the data is available after a while for debugging. The hour was very arbitrary.

If this causes big problems we can drop it. Some information about entering/leaving workers is also stored in the `all` topic but the worker specific topic is more granular.

Is there a problem with just canceling this task?
